crates/ruff_python_formatter/src/expression/expr_attribute.rs

```diff
@@ -88,55 +83,54 @@ impl FormatNodeRule<ExprAttribute> for FormatExprAttribute {
-                //         entry__pub_date__year=2008,
-                //     )
-                // )
-                // ```
-                write!(
-                    f,
-                    [
-                        (!leading_attribute_comments.is_empty()).then_some(hard_line_break()),
-                        leading_comments(leading_attribute_comments),
-                        text("."),
-                        trailing_comments(trailing_dot_comments),
-                        attr.format()
-                    ]
-                )
+            // Identify dangling comments before and after the dot:
+            // ```python
+            // (
+            //      (
+            //          a
+            //      )  # `before_dot_end_of_line`
+            //      # `before_dot_own_line`
+            //      .  # `after_dot_end_of_line`
+            //      # `after_dot_own_line`
+            //      b
+            // )
+            // ```
+            let comments = f.context().comments().clone();
+            let dangling = comments.dangling(item);
+            let (before_dot, after_dot) = if dangling.is_empty() {
+                (dangling, dangling)
             } else {
-                // Regular style
-                // ```python
-                // blogs2 = Blog.objects.filter(
-                //     entry__headline__contains="Lennon",
-                // ).filter(
-                //     entry__pub_date__year=2008,
-                // )
-                // ```
-                write!(
-                    f,
-                    [
-                        text("."),
-                        trailing_comments(trailing_dot_comments),
-                        (!leading_attribute_comments.is_empty()).then_some(hard_line_break()),
-                        leading_comments(leading_attribute_comments),
-                        attr.format()
-                    ]
+                let dot_token = find_only_token_in_range(
+                    TextRange::new(item.value.end(), item.attr.start()),
+                    SimpleTokenKind::Dot,
+                    f.context().source(),
+                );
+                dangling.split_at(
+                    dangling.partition_point(|comment| comment.start() < dot_token.start()),
                 )
-            }
+            };
+
+            let (before_dot_end_of_line, before_dot_own_line) = before_dot.split_at(
+                before_dot.partition_point(|comment| comment.line_position().is_end_of_line()),
+            );
+
+            let (after_dot_end_of_line, after_dot_own_line) = after_dot.split_at(
+                after_dot.partition_point(|comment| comment.line_position().is_end_of_line()),
+            );
+
+            write!(
+                f,
+                [
+                    trailing_comments(before_dot_end_of_line),
+                    (!before_dot.is_empty()).then_some(hard_line_break()),
+                    dangling_comments(before_dot_own_line),
```

Comment:
Is the writing of the hard line break still necessary after your `dangling_comments` change?

Reply:
No, good eye! But that's downstream of this, it's removed in the next PR.